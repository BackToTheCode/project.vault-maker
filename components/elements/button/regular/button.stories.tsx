import React from 'react';
import marked from 'marked';
import { Button } from '../regular';
import { checkA11y } from '@storybook/addon-a11y';
import { withKnobs, select } from '@storybook/addon-knobs';

export default {
  title: 'Elements|Button',
  component: Button,
  parameters: {
    info: {
      inline: true,
      header: false,
      text: marked(require('../button.stories.md').default)
    }
  },
  decorators: [checkA11y]
};

export const Primary = () => <Button variant={'primary'}>Go</Button>;

export const Secondary = () => <Button variant={'secondary'}>Go here</Button>;

export const Outline = () => <Button variant={'outline'}>Go back</Button>;

export const Text = () => <Button variant={'text'}>Click me</Button>;

export const WithKnobs = () => (
  <Button
    variant={select('Variant', ['primary', 'secondary', 'outline'], 'primary')}
  >
    Change me
  </Button>
);
WithKnobs.story = {
  decorators: [withKnobs]
};
