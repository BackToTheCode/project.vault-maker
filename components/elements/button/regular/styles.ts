const styles = {
    button:  {
        display: 'block',
        fontWeight: 'body',
        fontFamily: 'body',
        borderStyle: 'solid',
        borderWidth: 2,
        fontSize: 2,
        cursor: 'pointer',
        py: 4,
        px: 6,
        transition: '0.2s all',
        '&:hover': {
          transform: 'translateY(-1px)'
        },
        '&:focus': {
          outline: 0
        }
      },
  disabled: {
    cursor: 'default',
    opacity: '0.6',
    filter: 'grayscale(1)'
  }
};

export default styles;
