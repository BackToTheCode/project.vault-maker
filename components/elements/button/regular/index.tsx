/** @jsx jsx */
import { jsx } from '@emotion/core';
import React, { FC } from 'react';
import { Button as BaseButton } from 'rebass';
import styles from './styles';

export interface ButtonProps {
  /**
   * Simple click handler
   */
  onClick?: (e: any) => void;
  variant: string;
  isDisabled?: boolean;
  children?: React.ReactNode;
  sx?: any;
  icon?: string;
  /** Buttons background color */
  color?: 'blue' | 'green';
}


export const Button: FC<ButtonProps> = ({
  onClick,
  variant,
  isDisabled,
  sx,
  children
}) => {
  let mergedStyle = styles.button;
  if (isDisabled) {
    mergedStyle = { ...mergedStyle, ...styles.disabled };
  }

  return (
    <BaseButton
      onClick={onClick}
      sx={{ ...mergedStyle, ...sx }}
      variant={variant}
    >
      {children}
    </BaseButton>
  );
};

Button.defaultProps = {
  variant: 'primary'
};

Button.displayName = 'Button';