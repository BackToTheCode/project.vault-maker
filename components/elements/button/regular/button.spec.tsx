import { fireEvent, render } from '@testing-library/react';
import { ThemeProvider } from 'emotion-theming';
import React from 'react';
import appTheme from '../../../../styles/theme';
import { Button } from '../regular';

describe('<Button /> spec', () => {
  test('it renders the component', () => {
    // Arrange
    const buttonText = 'Click me';

    // Act
    const { getByText } = render(<Button>{buttonText}</Button>);

    // Assert
    expect(getByText(buttonText)).toMatchSnapshot();
  });

  test('that click gets handled', () => {
    // Arrange
    const buttonText = "Click me";
    jest.spyOn(window, 'alert').mockImplementation(() => {});
    const clickHandler = () => window.alert("Alerting you");
    
    // Act
    const { getByText } = render(<Button onClick={clickHandler}>{buttonText}</Button>);
    fireEvent.click(getByText(buttonText));

    // Assert
    expect(window.alert).toHaveBeenCalledTimes(1);
  });

  test('that <Button /> has correct background', () => {
    // Arrange
    const buttonText = 'Click me';
    const variant = "secondary";

    // Act 
    const { getByText } = render(<ThemeProvider theme={appTheme}><Button variant={variant}>{buttonText}</Button></ThemeProvider>);
    const button = getByText(buttonText);
    const style = window.getComputedStyle(button);

    // Assert
    expect(style.backgroundColor).toBe(appTheme.colors.secondary);
  });
});
