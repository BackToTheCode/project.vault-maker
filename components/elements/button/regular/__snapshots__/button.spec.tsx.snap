// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> spec it renders the component 1`] = `
<button
  class="css-zr4f50"
>
  Click me
</button>
`;
